features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/metric/SessionEndedMetric.kt

```diff
@@ -0,0 +1,221 @@
+            return false
+        }
+        val trackedView = trackedViewMap[rumViewEvent.view.id] ?: TrackedView(
+            viewUrl = rumViewEvent.view.url,
+            startMs = rumViewEvent.date,
+            durationNs = rumViewEvent.view.timeSpent
+        )
+
+        trackedViewMap[rumViewEvent.view.id] = trackedView.copy(durationNs = rumViewEvent.view.timeSpent)
+        if (firstTrackedView == null) {
+            firstTrackedView = trackedView
+        }
+        lastTrackedView = trackedView
+        return true
+    }
+
+    fun onErrorTracked(sdkErrorKind: String) {
+        trackedErrorMap[sdkErrorKind] = (trackedErrorMap[sdkErrorKind] ?: 0) + 1
+    }
+
+    fun onSessionStopped() {
+        wasStopped = true
+    }
+
+    fun toMetricAttributes(): Map<String, Any?> {
+        return resolveAttributes()
+    }
+
+    private fun resolveAttributes(): Map<String, Any?> {
+        return mapOf(
+            METRIC_TYPE_KEY to METRIC_TYPE_VALUE,
+            RSE_KEY to resolveRseAttributes()
+        )
+    }
+
+    private fun calculateDuration(): Long {
+        return lastTrackedView?.let { last ->
+            firstTrackedView?.let { first ->
+                last.startMs + last.durationNs - first.startMs
+            }
+        } ?: 0L
+    }
+
+    private fun resolveRseAttributes(): Map<String, Any?> {
+        return mapOf(
+            PROCESS_TYPE_KEY to PROCESS_TYPE_VALUE,
+            PRECONDITION_KEY to precondition,
+            DURATION_KEY to calculateDuration(),
+            WAS_STOPPED_KEY to wasStopped,
+            VIEW_COUNTS_KEY to resolveViewCountsAttributes(),
+            SDK_ERRORS_COUNT_KEY to resolveSDKErrorsCountAttributes()
+        )
+    }
+
+    private fun resolveViewCountsAttributes(): Map<String, Any?> {
+        return mapOf(
+            VIEW_COUNTS_TOTAL_KEY
+                to trackedViewMap.size,
+            VIEW_COUNTS_BG_KEY
+                to trackedViewMap.values.count { it.viewUrl == RumViewManagerScope.RUM_BACKGROUND_VIEW_URL },
+            VIEW_COUNTS_APP_LAUNCH_KEY
+                to trackedViewMap.values.count { it.viewUrl == RumViewManagerScope.RUM_APP_LAUNCH_VIEW_URL }
+        )
+    }
+
+    private fun resolveSDKErrorsCountAttributes(): Map<String, Any?> {
+        return mapOf(
+            SDK_ERRORS_COUNT_TOTAL_KEY to trackedErrorMap.values.sumOf {
+                it
+            },
+            SDK_ERRORS_COUNT_BY_KIND_KEY to resolveTop5ErrorsByKind()
+        )
+    }
+
+    private fun resolveTop5ErrorsByKind(): Map<String, Int> {
+        val map = mutableMapOf<String, Int>()
+        val limit = TOP_ERROR_LIMIT.coerceAtMost(trackedErrorMap.size)
+        trackedErrorMap.map { Pair(it.key, it.value) }.sortedByDescending {
+            it.second
+        }.subList(0, limit).forEach {
+            map[it.first.escapeNonAlphanumericCharacters()] = it.second
+        }
+        return map
+    }
+
+    private fun String.escapeNonAlphanumericCharacters(): String {
```

Comment:
nit: usually we are adding extension functions when it's being shared between multiple classes. Using a simple private function would be better here and more optimised.

Reply:
updated.